platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -264,6 +264,62 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     }
 }
 
+#pragma mark Offline merge methods
+
+- (void)addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    if (![fileManager isWritableFileAtPath:filePath]) {
+        [NSException raise:NSInvalidArgumentException format:@"The file path must be writable"];
+    }
+    
+    __weak MGLOfflineStorage *weakSelf = self;
+    [self _addContentsOfFile:filePath withCompletionHandler:^(NSError * _Nullable error) {
+        if (error && completion) {
+            completion(error);
+        } else {
+            MGLOfflineStorage *strongSelf = weakSelf;
+            [strongSelf getPacksWithCompletionHandler:^(NSArray<MGLOfflinePack *> *packs, NSError * _Nullable error) {
+                for (MGLOfflinePack *pack in strongSelf.packs) {
+                    [pack invalidate];
+                }
+                strongSelf.packs = [packs mutableCopy];
+                if (completion) {
+                    completion(error);
+                }
+            }];
+        }
+        
+    }];
+}
+
+- (void)addContentsOfURL:(NSURL *)fileURL withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    [self addContentsOfFile:fileURL.absoluteString withCompletionHandler:completion];
```

Comment:
`-[NSURL absoluteString]` returns a string beginning with `file:///`, not a file path. Instead, check that the `scheme` is `file` (raising an exception otherwise) and pass in `path`.

Add a test of this method that passes in a valid file URL.